subsys/usb/device_next/class/usbd_uvc.c

```diff
@@ -0,0 +1,2079 @@
+	struct k_mutex mutex;
+	/* Zero Length packet used to reset a stream when restarted */
+	struct net_buf zlp;
+};
+
+struct uvc_config {
+	/* Storage for the various descriptors available */
+	struct uvc_desc *desc;
+	/* Class context used by the USB device stack */
+	struct usbd_class_data *c_data;
+	/* Array of pointers to descriptors sent to the USB device stack and the host */
+	struct usb_desc_header **fs_desc;
+	struct usb_desc_header **hs_desc;
+};
+
+/* Specialized version of UDC net_buf metadata with extra fields */
+struct uvc_buf_info {
+	/* Regular UDC buf info so that it can be passed to USBD directly */
+	struct udc_buf_info udc;
+	/* Extra field at the end */
+	struct video_buffer *vbuf;
+	/* UVC stream this buffer belongs to */
+	const struct device *stream;
+} __packed;
+
+/* Mapping between UVC controls and Video controls */
+struct uvc_control_map {
+	uint32_t param;
+	enum uvc_op op;
+	uint8_t selector;
+	uint8_t size;
+	uint8_t bit;
+};
+
+struct uvc_guid_quirk {
+	uint32_t fourcc;
+	/* GUIDs are 16-bytes long, with the first four bytes being the Four Character Code of the
+	 * format and the rest constant, except for some exceptions listed in this table.
+	 */
+	uint8_t guid[16];
+};
+
+NET_BUF_POOL_VAR_DEFINE(uvc_buf_pool, DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) * 16,
+			512 * DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) * 16,
+			sizeof(struct uvc_buf_info), NULL);
+
+static void uvc_flush_queue(const struct device *dev);
+
+/* UVC public API */
+
+void uvc_set_video_dev(const struct device *const dev, const struct device *const video_dev)
+{
+	struct uvc_data *data = dev->data;
+
+	data->video_dev = video_dev;
+}
+
+/* UVC helper functions */
+
+static const struct uvc_guid_quirk uvc_guid_quirks[] = {
+	{VIDEO_PIX_FMT_YUYV,  UVC_FORMAT_GUID("YUY2")},
+};
+
+static void uvc_fourcc_to_guid(uint8_t guid[16], uint32_t fourcc)
+{
+	for (int i = 0; i < ARRAY_SIZE(uvc_guid_quirks); i++) {
+		if (uvc_guid_quirks[i].fourcc == fourcc) {
+			memcpy(guid, uvc_guid_quirks[i].guid, 16);
+			return;
+		}
+	}
+	fourcc = sys_cpu_to_le32(fourcc);
+	memcpy(guid, UVC_FORMAT_GUID("...."), 16);
+	memcpy(guid, &fourcc, 4);
+}
+
+static uint32_t uvc_guid_to_fourcc(uint8_t guid[16])
+{
+	uint32_t fourcc;
+
+	for (int i = 0; i < ARRAY_SIZE(uvc_guid_quirks); i++) {
+		if (memcmp(guid, uvc_guid_quirks[i].guid, 16) == 0) {
+			return uvc_guid_quirks[i].fourcc;
+		}
+	}
+
+	memcpy(&fourcc, guid, 4);
+
+	return sys_le32_to_cpu(fourcc);
+}
+
+/* UVC control handling */
+
+static const struct uvc_control_map uvc_control_map_ct[] = {
+	{
+		.size = 1,
+		.bit = 1,
+		.selector = UVC_CT_AE_MODE_CONTROL,
+		.op = UVC_OP_VC_FIXED,
+		.param = BIT(0)
+	},
+	{
+		.size = 1,
+		.bit = 2,
+		.selector = UVC_CT_AE_PRIORITY_CONTROL,
+		.op = UVC_OP_VC_FIXED,
+		.param = 0
+	},
+	{
+		.size = 4,
+		.bit = 3,
+		.selector = UVC_CT_EXPOSURE_TIME_ABS_CONTROL,
+		.op = UVC_OP_VC_INT,
+		.param = VIDEO_CID_EXPOSURE
```

Comment:
speaking of windows in your last comment... doesn't Windows also like to "not follow" the specification where it doesn't use 100us units for the exposure time absolute like Linux and Mac OS do, and sends it over in log2 value.

I know that in the past when I had to implement a 6 class USB composite device superspeed plus camera interface many years ago, and IIRC, Windows was kind of a nightmare to deal with 😠 , and did a not fun workaround to get users happy and put an extension unit in there as a quick bandaid. I'm not quite sure what good solution here would be... but maybe it's something to think about...

Reply:
Exposure being much more sensitive in the low values, I suppose they wanted to improve user experience of all the naive implementations exposing a linear exposure slider... Preventing to write correct software at the same time.

Currently, UVC does not really have a mapping between the controls and their integration. This is an arbitrary value passed to the sensor register directly, happily overflowing the max.

When https://github.com/zephyrproject-rtos/zephyr/pull/82158 is introduced, this would be the opportunitty to specify focus on how the UVC controls and Zephyr Video controls are mapped to each other, allowing some customization here.

Much grateful for this insight! This is the opportunity to put a long-term fix/workaround for these issues.

Reply:
Some proposal that could also help with such quirk and others:
- https://github.com/zephyrproject-rtos/zephyr/issues/86943